tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -310,11 +311,19 @@ void GatherGlobalsWithInitializers(
       // Add if user-defined static or groupshared global with initializer.
       if (VD->hasInit() && VD->hasGlobalStorage() &&
           (VD->getStorageClass() == SC_Static ||
-           VD->hasAttr<HLSLGroupSharedAttr>()))
-        GlobalsWithInit.push_back(VD);
+           VD->hasAttr<HLSLGroupSharedAttr>())) {
+        // Place subobjects in a separate collection.
+        QualType QT = VD->getType();
+        if (const RecordType *RT = QT->getAs<RecordType>()) {
+          RecordDecl *RD = RT->getDecl();
+          if (RD->hasAttr<HLSLSubObjectAttr>())
+            SubObjects.push_back(VD);
```

Comment:
If `RD` lacks `HLSLSubObjectAttr`, VD will not be added to `GlobalsWithInit`.  See my suggested change.  It uses `continue;`, instead of an `else` on the first branch.